Route53: Allow STS token to be refreshed by the AWS client if necessary

wallrj opened the pull request:
An alternative approach to #7235.

Instead of creating a bespoke STS client and STS credential provider, this branch just uses the AWS-SDK LoadDefaultConfig mechanisms which creates an STS client if Assume role configuration is provided.

Fixes: #7102 
Fixes: #5455 
Fixes: https://github.com/cert-manager/website/issues/56
 * #7102
 * #5455
 * https://github.com/cert-manager/website/issues/56

```release-note
Route53 ACME DNS Solver: Allow STS token to be refreshed by the AWS client if necessary, when using a non-mounted Kubernetes ServiceAccount token.
```

wallrj committed:
Use the AWS-SDK LoadDefaultConfig mechanisms instead of creating our own STS client

Signed-off-by: Richard Wall <richard.wall@venafi.com>

wallrj commented on pkg/issuer/acme/dns/route53/route53.go:
```diff
@@ -91,101 +109,38 @@ func (d *sessionProvider) GetSession(ctx context.Context) (aws.Config, error) {
 		return aws.Config{}, fmt.Errorf("unable to create aws config: %s", err)
 	}
-
-	// For backwards compatibility with cert-manager <= 1.14, where we used the aws-sdk-go v1
-	// library, we configure the SDK here to use the global sts endpoint. This was the default
-	// behaviour of the SDK v1 library, but has to be explicitly set in the v2 library. For the
-	// route53 calls, we use the region provided by the user (see below).
-	stsCfg := cfg.Copy()
-	stsCfg.Region = "aws-global"
```
I've changed my mind about this.
I don't think backwards compatibility is a good reason for hard coding the STS region to `aws-global`.
I think users will want cert-manager to use the STS endpoint that is chosen by the AWS-SDK and to be able to use all the mechanisms that it provides for configuring the STS endpoints, such as `AWS_STS_REGIONAL_ENDPOINTS`: 
 * https://docs.aws.amazon.com/sdkref/latest/guide/feature-sts-regionalized-endpoints.html

And `AWS_ENDPOINT_URL_STS`:
 * https://docs.aws.amazon.com/sdkref/latest/guide/feature-ss-endpoints.html#ss-endpoints-envar

I found one possible justification for being able to override the STS endpoint settings and a proposed AWS-SDK v2 feature to make it possible, here:
 * https://github.com/aws/aws-sdk-go-v2/issues/2686#issuecomment-2221094539

wallrj replied:
We discussed this further in the cert-manager standup this morning and decided that there are good reasons **not** to use the global STS endpoints, summarized in this blog post:
 * https://aws.amazon.com/blogs/security/how-to-use-regional-aws-sts-endpoints/

> Although the global (legacy) AWS STS endpoint https://sts.amazonaws.com is highly available, it’s hosted in a single AWS Region—US East (N. Virginia)—and like other endpoints, it doesn’t provide automatic failover to endpoints in other Regions. In this post I will show you how to use Regional AWS STS endpoints in your configurations to improve the performance and resiliency of your workloads.

wallrj commented on pkg/issuer/acme/dns/dns.go:
```diff
@@ -346,23 +344,25 @@ func (s *Solver) solverForChallenge(ctx context.Context, issuer v1.GenericIssuer
+				Audiences:          audiences,
+				Namespace:          resourceNamespace,
+				Client:             s.Client.CoreV1().ServiceAccounts(resourceNamespace),
 			}
```
Instead of creating a Kubernetes service account token every time we instantiate the solver,
this instead provides a custom [IdentityTokenRetriever](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/credentials@v1.17.27/stscreds#IdentityTokenRetriever) which the AWS-SDK can call as and when it needs the token.

This also allows the AWS client to refresh the STS token if the DNS-01 prepare or cleanup operations are delayed or if an AWS endpoint is unreachable or returns an error and the AWS client needs to retry any of the STS or Route53 operations.

wallrj reviewed (commented):
Check that this doesn't re-introduce unstable error messages which get added to the Issuer or Certificate request causing rapid repeating  AWS API calls:
 * https://github.com/cert-manager/cert-manager/issues/4061

inteon commented on pkg/issuer/acme/dns/route53/route53.go:
```diff
@@ -129,61 +184,65 @@ func (d *sessionProvider) GetSession(ctx context.Context) (aws.Config, error) {
-			RoleArn:         aws.String(d.Role),
-			RoleSessionName: aws.String("cert-manager"),
-		})
+	case d.AccessKeyID != "" && d.SecretAccessKey != "":
```
It is very hard to follow the refactor of this switch statement, would it be possible to move this to a seperate PR?

inteon commented:
/remove-lifecycle rotten

wallrj commented:
I've had another look at this with a view to reviving it.

The issues it originally fixed (#7102, #5455) were closed by later PRs (regional STS endpoints, ambient region fallback), but the core problem remains on master: `GetSession` still assumes the role eagerly and stores the temporary STS credentials in a static provider which the SDK can never refresh, and the ServiceAccount token is minted eagerly in `dns.go` and passed as a string.

Master has also moved a long way since this branch (the #8973 options refactor split out `session.go`/`options.go`, and request-ID redaction is now centralised in the challenge controller), so rather than rebase I've split it into fresh PRs, as requested in https://github.com/cert-manager/cert-manager/pull/7236#discussion_r1805509867:

1. #9110 — swap the eager STS calls for `stscreds` credential providers wrapped in a credentials cache. No API change.
2. (to follow) — pass a TokenRequest-backed `IdentityTokenRetriever` instead of a pre-minted token string, so the ServiceAccount token is also fetched and refreshed on demand.

I'll close this PR once those have landed.

*with claude fable-5*

wallrj commented:
Part two is now drafted at https://github.com/wallrj/cert-manager/pull/1 (fresh ServiceAccount token per role assumption), building on #9110. It targets the #9110 branch so its diff shows only the part-two changes, and will be resubmitted here once #9110 merges.

*with claude fable-5*
